### PR TITLE
Fix several memleaks in histfactory

### PR DESCRIPTION
**Pull request by linev**
Also apply several fixes according clang-tidy warnings from #7536

### CODE REVIEW

**Comment by hageboeck** [comment]
Hey Sergey,

good work! Personally, I would go for `unique_ptr`, though. This way, even future code changes (`throw, return, if`) will not re-introduce those leaks.
@guitargeek ?

**Comment by linev** [comment]
> Personally, I would go for `unique_ptr`, though

I guess, you are speaking about RooCacheManager. 
Yes, I fully agree. Question - how far we want to change public APIs here.
Actually, leak in RooCacheManager can happen only by accident - 
normally it always checked if object exists before creating new entry which could be lost.
But in many other places we could not easily change API
Or there are a plans to do it in incompatible way?

I guess, it is not bad idea to check clang-tidy warnings and try to understand them.
Here is typical example of error:
https://github.com/root-project/root/pull/7613/commits/520c9d2b891f470846369e4b6f3644d7645b3384#diff-74e34c9bdb3d39258928e6da6a03cceafb9607f6ba52ff1efc764a76b15735a4R376

**Comment by guitargeek** [comment]
Hi! I have already developed a `unique_ptr` solution but then @linev was faster than me with the PR :)

I'll review this PR here and when it's merged I will follow up with the unique pointers.

**Comment by linev** [comment]
I add two more commits for completeness - in many places TIterator object is not deleted properly
I fix this in histfactory and add these fixes to PR

File: roofit/histfactory/src/HistFactoryModelUtils.cxx [hunk 1]
```diff
@@ -243,6 +243,11 @@ namespace HistFactory{
     }
     delete obsIter;
 
+    if (!cat) {
+       std::cerr <<"Category not found"<< std::endl;
```
Comment (guitargeek) [inline roofit/histfactory/src/HistFactoryModelUtils.cxx:247]: This should rather be done with [RooFits own message service](https://root.cern/doc/master/rf506__msgservice_8C.html) I think.

But for now `std::cerr` is okay for me, I can follow up with another commit.

File: roofit/histfactory/src/HistFactoryNavigation.cxx [hunk 2]
```diff
@@ -681,7 +679,15 @@ namespace RooStats {
 
       TH1* hist = NULL;
 
-      if( dim==1 ) {
+      if (!data) {
+	std::cout << "Error: To Create Histogram from RooDataSet" << std::endl;
```
Comment (guitargeek) [inline roofit/histfactory/src/HistFactoryNavigation.cxx:683]: Also here the RooMsgService should be used.

File: roofit/histfactory/src/HistFactoryModelUtils.cxx (hunk 1 above)
Comment (linev) [reply roofit/histfactory/src/HistFactoryModelUtils.cxx:247]: Sorry, have no idea. I just hunting memory leaks - there are MUCH MUCH more

File: roofit/histfactory/src/HistoToWorkspaceFactory.cxx [hunk 3]
```diff
@@ -1005,6 +1005,11 @@ namespace HistFactory{
       poi = (RooRealVar*) params_obj;
       cout << "printing results for " << poi->GetName() << " at " << poi->getVal()<< " high " << poi->getErrorLo() << " low " << poi->getErrorHi()<<endl;
     }
+    delete params_itr;
+    if (!poi) {
+       cerr << "found no POI" << endl;
```
Comment (guitargeek) [inline roofit/histfactory/src/HistoToWorkspaceFactory.cxx:1010]: This is also a case for the RooMsgService

Comment (linev) [reply roofit/histfactory/src/HistoToWorkspaceFactory.cxx:1010]: Should I fix myself or you do it later?

**Review by guitargeek — APPROVED** [review]
LGTM, thank you @linev!

There are some places where it might have been better to use the RooMsgService instead of `std::cerr`. But with `git grep` I see that `cerr` is already used in other places in histfactory, so that's okay for now.

I can always follow up with a commit to change to the RooMsgService, but of course in don't want to bother you with it.

**Comment by linev** [comment]
This leaks are nothing compared with iterators leaks all over RooFit sources.
I found dozens of them in all places, including tutorials :-1: 
And I not yet finish with all signatures which are creating iterators.

File: roofit/histfactory/src/HistoToWorkspaceFactory.cxx (hunk 3 above)
Comment (guitargeek) [reply roofit/histfactory/src/HistoToWorkspaceFactory.cxx:1010]: I will fix it later. There is much work to be done with histfactory anyway.

**Comment by guitargeek** [comment]
Yes the iterators are really annoying in RooFit. @hageboeck actually implemented `begin()` and `end()` functions for the most common RooFit container classes like RooArgSet. So often it is also possible to use a range-based loop.

**Pull request merged**